### DISCUSSION TITLE
How to update metadata?

### DISCUSSION THREAD

**Discussion by synox** [RhetTbull/osxphotos#505] [General]
My wife and I have separate Photos Library. I would like to transfer her likes, and mark my photos - if they exists. 

The metadata export is easy with osxphtos. But how can I update metadata (likes, rating) of Photos?

**Comment by RhetTbull** [comment]
@synox if you know Python, you could write a script to do this using my other Python package [photoscript](https://github.com/RhetTbull/PhotoScript).  photoscript allows you to interact with Photos using Python.  The most difficult part will be finding the matching photos from your wife's library and matching those to the photos in your library.  While each photo in the library has a unique identification (called a UUID), this is unique to each particular library not to the photo itself so the same photo in her library will have a different UUID than the one in yours.  If the photos we copied directly from the original source to both libraries then you could compare hash of the photos (see [hashlib](https://docs.python.org/3/library/hashlib.html)) to find matching images.  It's possible the hashes in her library don't match the hashes in yours if the originals were modified before being copied to one of the libraries.  In that case you'd have to use something like a [perceptual hash](https://github.com/JohannesBuchner/imagehash).

The basic script would look like this:

- use osxphotos to go through  all her images, store a hash of each original image
- use osxphotos to go through all your images, store a hash of each original image
- find matching hashes then get the data from the photo in her library
- update the metadata in your library using photoscript

To get you started, here's the basics for going through the library with osxphotos:

```python
import osxphotos

photosdb = osxphotos.PhotosDB("/path/to/first/library")

for photo in photosdb.photos():
    print(photo.path, photo.uuid)
    # hash the photo at the path
    # metadata can be accessed in this way
    print(photo.favorite, photo.description)
```

And to update the metadata:
```python
import photoscript

photoslibrary = photoscript.PhotosLibrary()
photoslibrary.open("/path/to/library")
photo = photoscript.Photo("uuid value")
photo.favorite = True
photo.description = "New Description"
```

**Reply by synox** [reply]
Thank you for your help getting started. I think this might work. The original hashes should be the same, or at least the originalFilename/date combination should be unique. 


Edit: After having issues with your script, I tried another example of `photoscript` and got the following error. 
`applescript.ScriptError: Not authorized to send Apple events to Photos. (-1743) app='Photos' range=6228-6234`

I solved it by running it from Terminal instead of IntelliJ. On my machine, Terminal already has automation permission for Photos.

**Comment by RhetTbull** [comment]
> applescript.ScriptError: Not authorized to send Apple events to Photos. (-1743) app='Photos' range=6228-6234

Yes, due to how locked down Apple has made macOS, the app running the script needs to request permission to use AppleScript.  Some editors will do this for you but apparently IntelliJ does not.
